src/models/disassemblyoutput.cpp

```diff
@@ -119,6 +119,52 @@ QString findBinaryForSymbol(const QStringList& debugPaths, const QStringList& ex
 
     return {};
 }
+
+int longestCommonLengthFromEnd(const QString& a, const QString& b)
+{
+    const auto lengthA = a.length();
+    const auto lengthB = b.length();
+    int shortestLength = std::min(lengthA, lengthB);
+    int inCommon = 0;
+
+    for (int i = 1; i < shortestLength; i++) {
+        if (a[lengthA - i] != b[lengthB - i])
+            break;
+        inCommon++;
+    }
+
+    return inCommon;
+}
+}
+
+// not in an anonymous namespace so we can test this
+QString findSourceCodeFile(const QString& originalPath, const QStringList& sourceCodePaths)
+{
+    if (sourceCodePaths.isEmpty())
+        return originalPath;
+
+    int longestMatch = 0;
+    QString candidate;
+
+    QFileInfo info(originalPath);
+
+    for (const auto& path : sourceCodePaths) {
+        // we want the longest matching path since some projects are organized like this:
+        // - liba/lib.c
+        // - libb/lib.c
+
+        QDirIterator it(path, {info.fileName()}, QDir::NoFilter, QDirIterator::Subdirectories);
```

Comment:
if someone adds `/` to the source code paths then this will become _really_ slow as we'll be basically looping the whole file system for every source code file we are looking up?

instead, we should:

- first lookup the originalPath directly
- then fallback to the originalPath in the sysroot
- then loop over the source paths in this manner:

given a path `/foo/bar/file.c` and a source path `$SP` you'd start lookup:

```
$SP/foo/bar/file.c
$SP/bar/file.c
$SP/file.c
```

i.e. pop the dirs from the front of the source path until only the filename remains and attach that to the source dir.

I think ideally the loop should also be such that we first lookup the full paths in all source dirs, then pop the dir from front of the source path and try again all source dirs. that way the most specific path in any of the source dirs wins, which sounds correct to me?